orion/primitives/tadgan.py

```diff
@@ -82,244 +142,317 @@ def __getstate__(self):
         self.critic_x_input_shape = self.target_shape
         self.critic_z_input_shape = self.latent_shape
 
-    def _build_tadgan(self, **kwargs):
+    def _build_tadgan(self, **kwargs) -> None:
         hyperparameters = self.hyperparameters.copy()
         hyperparameters.update(kwargs)
 
-        self.encoder = self._build_model(hyperparameters, self.layers_encoder,
-                                         self.encoder_input_shape)
-        self.generator = self._build_model(hyperparameters, self.layers_generator,
-                                           self.generator_input_shape)
-        self.critic_x = self._build_model(hyperparameters, self.layers_critic_x,
-                                          self.critic_x_input_shape)
-        self.critic_z = self._build_model(hyperparameters, self.layers_critic_z,
-                                          self.critic_z_input_shape)
-
-        self.generator.trainable = False
-        self.encoder.trainable = False
-
-        x = Input(shape=self.shape)
-        y = Input(shape=self.target_shape)
-        z = Input(shape=(self.latent_dim, 1))
+        # Models
+        self.encoder = self._build_model(
+            hyperparameters, self.layers_encoder, self.encoder_input_shape, name='encoder')
+        self.generator = self._build_model(
+            hyperparameters, self.layers_generator, self.generator_input_shape, name='generator')
+        self.critic_x = self._build_model(
+            hyperparameters, self.layers_critic_x, self.critic_x_input_shape, name='critic_x')
+        self.critic_z = self._build_model(
+            hyperparameters, self.layers_critic_z, self.critic_z_input_shape, name='critic_z')
+
+        # Optimizers
+        self.critic_x_optimizer = import_object(self.optimizer)(self.learning_rate)
+        self.critic_z_optimizer = import_object(self.optimizer)(self.learning_rate)
+        self.encoder_generator_optimizer = import_object(self.optimizer)(self.learning_rate)
+
+    def _format_losses(self, losses: list) -> dict:
+        """Format losses into dictionary mapping metric names to their current value."""
+        output = dict()
+        if self.detailed_losses:
+            for i in range(len(losses)):
+                for j in range(len(losses[i])):
+                    output[LOSS_NAMES[i][j]] = losses[i][j]
+        else:
+            for i in range(len(losses)):
+                output[LOSS_NAMES[i][0]] = losses[i][0]
+        return output
+
+    def compile(self, **kwargs):
+        super(TadGAN, self).compile(**kwargs)
+
+    @staticmethod
+    @tf.function
+    def _wasserstein_loss(y_true, y_pred):
+        return tf.reduce_mean(y_true * y_pred)
+
+    @staticmethod
+    @tf.function
+    def _gradient_penalty_loss(real, fake, critic: Model):
+        # Random weighted average to create interpolated signals.
+        batch_size = tf.shape(real)[0]
+        alpha = tf.random.uniform([batch_size, 1, 1], dtype=tf.float64)
```

Comment:
curious if there is a reason to calculate `batch_size` rather than directly using `self.batch_size`?

Reply:
This is for the edge case if the batch is less than `batch_size`, e.g. the leftover samples